

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>synbio &mdash; synbio 0.4.24 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Reference" href="ref/ref.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="#" class="icon icon-home"> synbio
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="ref/ref.html">Reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="ref/designs.html">Designs</a></li>
<li class="toctree-l2"><a class="reference internal" href="ref/protocols.html">Protocols</a></li>
<li class="toctree-l2"><a class="reference internal" href="ref/steps.html">Steps</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a><ul>
<li class="toctree-l2"><a class="reference internal" href="examples.html#assembly">Assembly</a><ul>
<li class="toctree-l3"><a class="reference internal" href="examples.html#gibson-assembly-with-primer-generation">Gibson Assembly with primer generation</a></li>
<li class="toctree-l3"><a class="reference internal" href="examples.html#cloning-via-restriction-digest-and-ligation">Cloning via restriction digest and ligation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="examples.html#protocols">Protocols</a><ul>
<li class="toctree-l3"><a class="reference internal" href="examples.html#gibson-assembly-of-a-plasmid-library">Gibson Assembly of a plasmid library</a></li>
<li class="toctree-l3"><a class="reference internal" href="examples.html#gibson-assembly-of-a-single-plasmid">Gibson Assembly of a single plasmid</a></li>
<li class="toctree-l3"><a class="reference internal" href="examples.html#golden-gate-assembly-of-a-single-plasmid">Golden Gate Assembly of a single plasmid</a></li>
<li class="toctree-l3"><a class="reference internal" href="examples.html#combinatorial-golden-gate-assembly-all-valid-plasmid-combinations">Combinatorial Golden Gate Assembly: all valid plasmid combinations</a></li>
<li class="toctree-l3"><a class="reference internal" href="examples.html#combinatorial-golden-gate-assembly-with-bins-all-plasmid-combinations-between-bins">Combinatorial Golden Gate Assembly with bins: all plasmid combinations between bins</a></li>
</ul>
</li>
</ul>
</li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">synbio</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="#">Docs</a> &raquo;</li>
        
      <li>synbio</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <!-- User defined GitHub URL -->
              <a href="https://github.com/Lattice-Automation/synbio" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="synbio">
<h1>synbio<a class="headerlink" href="#synbio" title="Permalink to this headline">¶</a></h1>
<p><code class="docutils literal notranslate"><span class="pre">synbio</span></code> is a library for designing and assembling DNA. Users can design plasmids or libraries and export multi-step build protocols. Input SeqRecords; output assembly SeqRecords, protocols, plate maps, and robotic picklists.</p>
<div class="section" id="installation">
<h2>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h2>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># with pip</span>
pip install synbio

<span class="c1"># OR with conda</span>
conda install -c jtimmons synbio
</pre></div>
</div>
</div>
<div class="section" id="models">
<h2>Models<a class="headerlink" href="#models" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">synbio</span></code> only expects the user to define their <code class="docutils literal notranslate"><span class="pre">Design</span></code> and <code class="docutils literal notranslate"><span class="pre">Protocol</span></code>. Several protocols are pre-defined.</p>
<div class="section" id="designs">
<h3>Designs<a class="headerlink" href="#designs" title="Permalink to this headline">¶</a></h3>
<p>All are in <code class="docutils literal notranslate"><span class="pre">synbio.designs</span></code>:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">Combinatorial</span></code> - list of SeqRecords to combinatorially anneal into all valid assemblies</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">CombinatorialBins</span></code> - list of bins of SeqRecords for combinatorial assembly between bins</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Plasmid</span></code> - single list of SeqRecords to combine into a plasmid</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">PlasmidLibrary</span></code> - list of list of SeqRecords to combine into plasmids</p></li>
</ul>
</div>
<div class="section" id="protocols">
<h3>Protocols<a class="headerlink" href="#protocols" title="Permalink to this headline">¶</a></h3>
<p>All are in <code class="docutils literal notranslate"><span class="pre">synbio.protocols</span></code>:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">Gibson</span></code> - Gibson assembly based on <a class="reference external" href="https://www.neb.com/protocols/2012/12/11/gibson-assembly-protocol-e5510">NEB’s E5510</a></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">GoldenGate</span></code> - Golden Gate assembly based on <a class="reference external" href="https://www.neb.com/products/e1601-neb-golden-gate-assembly-mix#Product%20Information">NEB’s E1601</a></p></li>
</ul>
</div>
</div>
<div class="section" id="example">
<h2>Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h2>
<p>In the example below, the user specifies a combinatorial library design. All SeqRecords are tested for circularization with other SeqRecords. New and valid plasmids are assembled.</p>
<p>Behind the scenes, <code class="docutils literal notranslate"><span class="pre">synbio</span></code> is filtering all combinations of SeqRecords from the design that will circularize into valid plasmids (via <a class="reference external" href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-015-0544-x/figures/1">circuits in a graph</a>). After running the <code class="docutils literal notranslate"><span class="pre">protocol</span></code>, users can export plate maps (<code class="docutils literal notranslate"><span class="pre">to_csv()</span></code>), composite plasmids (<code class="docutils literal notranslate"><span class="pre">to_fasta()</span></code>, <code class="docutils literal notranslate"><span class="pre">to_genbank()</span></code>), and assembly instructions (<code class="docutils literal notranslate"><span class="pre">to_txt()</span></code>, <code class="docutils literal notranslate"><span class="pre">to_picklists()</span></code>).</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="sd">&quot;&quot;&quot;Example of a Combinatorial Golden Gate assembly with human and robot output protocols.&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">os</span>

<span class="kn">from</span> <span class="nn">Bio.SeqIO</span> <span class="kn">import</span> <span class="n">parse</span>

<span class="kn">from</span> <span class="nn">synbio.designs</span> <span class="kn">import</span> <span class="n">Combinatorial</span>
<span class="kn">from</span> <span class="nn">synbio.protocols</span> <span class="kn">import</span> <span class="n">GoldenGate</span>

<span class="k">def</span> <span class="nf">read_all_records</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;Gather all SeqRecords from &quot;goldengate&quot; dir in examples.&quot;&quot;&quot;</span>

    <span class="n">GG_DIR</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">,</span> <span class="s2">&quot;examples&quot;</span><span class="p">,</span> <span class="s2">&quot;goldengate&quot;</span><span class="p">)</span>

    <span class="n">records</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="nb">file</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">GG_DIR</span><span class="p">):</span>
        <span class="k">if</span> <span class="nb">file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.gb&quot;</span><span class="p">):</span>
            <span class="n">records</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">parse</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">GG_DIR</span><span class="p">,</span> <span class="nb">file</span><span class="p">),</span> <span class="s2">&quot;genbank&quot;</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">records</span>

<span class="c1"># create a combinatorial library design from all valid combinations</span>
<span class="n">design</span> <span class="o">=</span> <span class="n">Combinatorial</span><span class="p">(</span><span class="n">read_all_records</span><span class="p">())</span>

<span class="c1"># create a protocol using Golden Gate as the sole composite step and run</span>
<span class="n">protocol</span> <span class="o">=</span> <span class="n">GoldenGate</span><span class="p">(</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;CombinatorialBins Golden Gate&quot;</span><span class="p">,</span>
    <span class="n">design</span><span class="o">=</span><span class="n">design</span><span class="p">,</span>
    <span class="n">include</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;KanR&quot;</span><span class="p">],</span>  <span class="c1"># only keep circularized plasmids with KanR</span>
    <span class="n">min_count</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>  <span class="c1"># only keep circularized plasmids from &gt;=5 SeqRecords</span>
<span class="p">)</span>

<span class="n">protocol</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>

<span class="c1"># export all the output plasmids to a multi-FASTA</span>
<span class="n">protocol</span><span class="o">.</span><span class="n">to_fasta</span><span class="p">(</span><span class="s2">&quot;plasmids.fasta&quot;</span><span class="p">)</span>

<span class="c1"># export plate layouts</span>
<span class="n">protocol</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;plates.csv&quot;</span><span class="p">)</span>

<span class="c1"># export human protocol</span>
<span class="n">protocol</span><span class="o">.</span><span class="n">to_txt</span><span class="p">(</span><span class="s2">&quot;protocol.txt&quot;</span><span class="p">)</span>

<span class="c1"># export a hamilton picklist</span>
<span class="n">protocol</span><span class="o">.</span><span class="n">to_picklists</span><span class="p">(</span><span class="s2">&quot;picklist&quot;</span><span class="p">,</span> <span class="n">platform</span><span class="o">=</span><span class="s2">&quot;hamilton&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>plasmids.fasta:</p>
<div class="highlight-txt notranslate"><div class="highlight"><pre><span></span>&gt;J23100_AB+B0032m_BC+C0012m_CD+B0015_DE+DVK_AE
GGAGTTGACGGCTAGCTCAGTCCTAGGTACAGTGCTAGCTACTAGAGTCACACAGGAAAG
TACTAAATGATGGTGAATGTGAAACCAGTAACGTTATACGATGTCGCAGAGTATGCCGGT
...
</pre></div>
</div>
<p>plates.csv:</p>
<div class="highlight-csv notranslate"><div class="highlight"><pre><span></span>Setup Wells with volumes (uL) shown:
Plate:1,1,2,3,4,5,6,7,8,9,10,11,12
A,B0015_DE(4),C0080_CD(18),R0010_AB(54),water(36)
B,B0015_DE(160),DVK_AE(160),cre_CD(18),water(156)
...
</pre></div>
</div>
<p>protocol.txt:</p>
<div class="highlight-txt notranslate"><div class="highlight"><pre><span></span>Combinatorial GoldenGate:
1. Setup PCR plate with (volumes) shown:
    1.1. Dilute plasmid DNA to 75 ng/µL in &#39;water&#39;
    1.2. Create &#39;assembly-mix&#39; from 1:1 T4 Ligase Buffer (10X) and NEB Golden Gate Assembly Mix
...
</pre></div>
</div>
<p>picklist.gwl:</p>
<div class="highlight-txt notranslate"><div class="highlight"><pre><span></span>A;Plate:2;;;15;;2.0;;;
D;Plate:3;;;80;;2.0;;;
W;;;;;;;;;
...
</pre></div>
</div>
</div>
<div class="section" id="alternatives">
<h2>Alternatives<a class="headerlink" href="#alternatives" title="Permalink to this headline">¶</a></h2>
<p>This is a non-exhaustive list. Contact me for a comparison of these libraries/platforms and <code class="docutils literal notranslate"><span class="pre">synbio</span></code>.</p>
<ul class="simple">
<li><p><a class="reference external" href="https://www.aquarium.bio/">Aquarium</a> is an extensive library/application for LIMS, protocol definition/execution, and workflow design. A lab operating system.</p></li>
<li><p><a class="reference external" href="https://github.com/autoprotocol/autoprotocol-python">Autoprotocol</a> is a specification standard for experiments in the life sciences.</p></li>
<li><p><a class="reference external" href="https://github.com/liaupm/BioBlocks">BioBricks</a> is a general focus, web-based editor for describing experiments in Biology.</p></li>
<li><p><a class="reference external" href="https://jbioleng.biomedcentral.com/articles/10.1186/1754-1611-4-13">Biocoder</a> is a C++ library with extensive protocol step definition capabilities.</p></li>
<li><p><a class="reference external" href="https://github.com/Edinburgh-Genome-Foundry/Plateo">Plateo</a> is a python library for planning, running and checking laboratory experiments. Great for parsing and exporting plates and picklists form multiple formats.</p></li>
<li><p><a class="reference external" href="https://github.com/BjornFJohansson/pydna">pydna</a> is a python DNA assembly simulation library with a human-readable description of clone and assembly strategies.</p></li>
</ul>
</div>
</div>
<div class="section" id="contents">
<h1>Contents<a class="headerlink" href="#contents" title="Permalink to this headline">¶</a></h1>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="ref/ref.html">Reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="ref/designs.html">Designs</a></li>
<li class="toctree-l2"><a class="reference internal" href="ref/protocols.html">Protocols</a></li>
<li class="toctree-l2"><a class="reference internal" href="ref/steps.html">Steps</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a><ul>
<li class="toctree-l2"><a class="reference internal" href="examples.html#assembly">Assembly</a><ul>
<li class="toctree-l3"><a class="reference internal" href="examples.html#gibson-assembly-with-primer-generation">Gibson Assembly with primer generation</a></li>
<li class="toctree-l3"><a class="reference internal" href="examples.html#cloning-via-restriction-digest-and-ligation">Cloning via restriction digest and ligation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="examples.html#protocols">Protocols</a><ul>
<li class="toctree-l3"><a class="reference internal" href="examples.html#gibson-assembly-of-a-plasmid-library">Gibson Assembly of a plasmid library</a></li>
<li class="toctree-l3"><a class="reference internal" href="examples.html#gibson-assembly-of-a-single-plasmid">Gibson Assembly of a single plasmid</a></li>
<li class="toctree-l3"><a class="reference internal" href="examples.html#golden-gate-assembly-of-a-single-plasmid">Golden Gate Assembly of a single plasmid</a></li>
<li class="toctree-l3"><a class="reference internal" href="examples.html#combinatorial-golden-gate-assembly-all-valid-plasmid-combinations">Combinatorial Golden Gate Assembly: all valid plasmid combinations</a></li>
<li class="toctree-l3"><a class="reference internal" href="examples.html#combinatorial-golden-gate-assembly-with-bins-all-plasmid-combinations-between-bins">Combinatorial Golden Gate Assembly with bins: all plasmid combinations between bins</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="ref/ref.html" class="btn btn-neutral float-right" title="Reference" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>